src/org/rascalmpl/library/String.rsc

```diff
@@ -680,4 +682,119 @@ or the indentation.
 * This function works fine if `indentation` is not spaces or tabs; but it does not make much sense. 
 }
 @javaClass{org.rascalmpl.library.Prelude}
-java str indent(str indentation, str content, bool indentFirstLine=false);
\ No newline at end of file
+java str indent(str indentation, str content, bool indentFirstLine=false);
+
+list[str] newLineCharacters = [
+    "\u000A", // LF
+    "\u000B", // VT
+    "\u000C", // FF
+    "\u000D", // CR
+    "\u000D\u000A", // CRLF
+    "\u0085", // NEL
+    "\u2028", // LS
+    "\u2029" // PS
+];
+
+@synopsis{Comparator to sort strings by length (ascending).}
+private bool bySize(str a, str b) = size(a) < size(b);
+
+@synopsis{Comparator to sort strings by relative position in a reference list.}
+private bool(str, str) byIndex(list[str] indices) {
+    return bool(str a, str b) {
+        return indexOf(indices, a) < indexOf(indices, b);
+    };
+}
+
+@synopsis{Determine the most-used newline character in a string.}
+str mostUsedNewline(str input, list[str] lineseps = newLineCharacters, str(list[str]) tieBreaker = getFirstFrom) {
+    linesepCounts = (nl: 0 | nl <- lineseps);
+    for (nl <- sort(lineseps, bySize)) {
+        int count = size(findAll(input, nl));
+        linesepCounts[nl] = count;
+        // subtract all occurrences of substrings of newline characters that we counted before
+        for (str snl <- substrings(nl), linesepCounts[snl]?) {
+            linesepCounts[snl] = linesepCounts[snl] - count;
+        }
+    }
+
+    byCount = invert(linesepCounts);
+    return tieBreaker(sort(byCount[max(domain(byCount))], byIndex(lineseps)));
+}
+
+@synopsis{Split a string to an indentation prefix and the remainder of the string.}
+tuple[str indentation, str rest] splitIndentation(/^<indentation:\s*><rest:.*>/)
+    = <indentation, rest>;
+
+str(str) indentSpacesAsTabs(int tabSize) {
```

Comment:
What's the reason that these two are higher-order functions?

Reply:
So you can do stuff like [this](https://github.com/usethesource/rascal-language-servers/pull/677/files#diff-1f4a2bdb22de60d7bf7b39aedcc6e36f04626d4287a5ac7b976dfb65aca5e5a2R91)
```formatted = perLine(formatted, indentSpacesAsTabs(opts.tabSize));```

Also, I figured constructing that replacement string was relatively expensive. This way, you only need to do it once.